### ISSUE TITLE
Add support for the s390x architecture

### ISSUE THREAD

**Issue by fradaham** [dotnet/runtime#34195]
As a team responsible for a self-developed integration platform implemented in asp.net core (currently 3.1) with the task of exposing IBM mainframe services, we have an interest to be able to run this platform as a linux container in the IBM zCX (zOS container extension) docker environment. But its a showstopper that there aren't any released linux images with .net (sdk and runtime) built for the s390x architecture used in IBM zCX. There are several other programming languages represented as linux docker containers for the s390x architecture, like python, java, ruby, golang, groovy, perl, just to mention some (see https://hub.docker.com/u/s390x/?page=1)).  Is it possible to build .net core (sdk and runtime (asp.net core)) for the s390x architecture in a linux container? There exists a debian s390x docker image (https://hub.docker.com/r/s390x/debian).

**Comment by jkotas** [comment]
https://github.com/mono/mono supports s390x . Now that Mono is being unified with .NET Core, the easiest way to make progress on s390x would be to enable s390x Mono in https://github.com/dotnet/runtime .

**Comment by MichaelErichsen** [comment]
This could have general interest. I have described a scenario that were interesting to us at "Running a hybrid application in Docker and zCX on ZD&T" [Part 1](https://www.linkedin.com/pulse/running-hybrid-application-docker-zcx-zdt-michael-erichsen/) and [Part 2](https://www.linkedin.com/pulse/building-hybrid-cicsdotnet-application-docker-zcx-zdt-erichsen/).

**Comment by jperezpbos** [comment]
Adding a quick comment from https://github.com/dotnet/core/issues/4917 for visibility and clarification.

Not really need to run as a container as long as we have .NET 5 for s390x for a Linux distribution we can deploy in different ways.  
Support for s390x (IBM Z / LinuxONE platform) is a great opportunity to take over mostly Java and Python development in s390x right now. Note that this platform is used for enterprise applications for the largest Financial Services organizations and a lot of FinTech, Blockchain, Data Science, and AI development.

**Comment by jperezpbos** [comment]
> https://github.com/mono/mono supports s390x . Now that Mono is being unified with .NET Core, the easiest way to make progress on s390x would be to enable s390x Mono in https://github.com/dotnet/runtime .

Looks like this was done here https://github.com/mono/mono/pull/19611 by @nealef

**Comment by neman** [comment]
Looks like it was not done since build is failing

![image](https://user-images.githubusercontent.com/350314/95865852-6f001100-0d67-11eb-9a57-f276f1504e6c.png)

**Comment by jperezpbos** [comment]
Thanks @neman.  Hi @nealef is this something you can fix?

**Comment by nealef** [comment]
Are you asking can dotnet/runtime be built for s390x? If so, I am working on it. If you are asking can mono be built from github.com/mono for s390x then yes.

**Comment by neman** [comment]
So can we expect that .net 5 will have support for s390x anytime soon?

**Comment by nealef** [comment]
Depends on a number of factors:
1. Getting time to do the work
2. Learning what needs to be done other than the fixes to build scripts and CMakeLists.txt
3. Having that work reviewed and accepted

**Comment by nealef** [comment]
We're making progress...

```
$ ~/dotNET/dotnet HelloWorld.dll
Hello World!
```

```
$ ~/dotNET/dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.100-rc.2.20479.15
 Commit:    da7dfa8840

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         linux-s390x
 Base Path:   /home/neale/dotNET/sdk/6.0.0/

Host (useful for support):
  Version: 6.0.0-dev
  Commit:  8a061afef9

.NET SDKs installed:
  6.0.0 [/home/neale/dotNET/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.0-rc.2.20475.17 [/home/neale/dotNET/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.0-rc.2.20475.5 [/home/neale/dotNET/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-dev [/home/neale/dotNET/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by ad-ops** [comment]
Is there any progress on this issue? Is there any better support in .Net 6?

**Comment by nealef** [comment]
Take a look in dotnet/runtimelab repo and the feature/s390x branch. You can cross-build dotnet from it. There are also fixes pending for msbuild and NuGet.Client that will enable the dotnet new and build operations. The dotnet command will run built apps without those fixes.

**Comment by fradaham** [comment]
Is this fully included now in the new .net 6.0 release? I guess so: https://devblogs.microsoft.com/dotnet/announcing-net-6/. But are there any official linux images including .net for the s390x architecture? I found one for RHEL...but would have preferred one published on docker hub.

**Comment by uweigand** [comment]
The only binary build of .NET on s390x available at this point is the one provided by Red Hat.  RPM packages are available with RHEL 8.5, and Red Hat also provides a set of UBI-based container images containing the .NET runtime & SDK:
https://catalog.redhat.com/software/containers/rhel8/dotnet-60/6182efa2be25a74c00923848?architecture=s390x

Does this work for you?

You can of course always build .NET yourself from source, but that is somewhat non-trivial, in particular since .NET is written in .NET, so you first need *some* .NET runtime available.  This can be handled either by starting with one of existing images (from Red Hat), or by performing a cross-build starting on another architecture like Intel.

**Comment by fradaham** [comment]
Ok, we have gotten hold of the RedHat image, and will start trying it out. But what if we, or anyone else, wants to build .NET for another linux image? Are there some instructions on how to do this, making it somewhat less non-trivial? Either the cross-build case, or starting from the RedHat image?

**Comment by tonyqus** [comment]
> Looks like it was not done since build is failing
> 
> ![image](https://user-images.githubusercontent.com/350314/95865852-6f001100-0d67-11eb-9a57-f276f1504e6c.png)

Where is this build status page? Is it open to public?

**Comment by jmertic** [comment]
If there is interest in the project in working on this port, it might be a good Open Mainframe Project Mentorship project...

https://www.openmainframeproject.org/projects/mentorship-program

Let me know if there is interest and I can help get this posted.

**Comment by uweigand** [comment]
> If there is interest in the project in working on this port, it might be a good Open Mainframe Project Mentorship project...
> 
> https://www.openmainframeproject.org/projects/mentorship-program
> 
> Let me know if there is interest and I can help get this posted.

Hi John, just to clarify: as mentioned above, we **have** ported .NET to s390x - all required changes are present in the upstream sources starting with .NET 6, and Red Hat provides binary images as RPMs (part of RHEL 8.5 and later) and as container images (linked above).

The only missing question is whether there is interest in someone providing binary builds in addition to what Red Hat is doing (e.g. to support other distros).   On other platforms, Microsoft are providing regular binary builds supporting multiple Linux distros, but they do not do so for the s390x architecture.

At this point, this is not so much a question of porting, but rather of setting up a CI infrastructure that regularly checks out the current sources, builds it for s390x (e.g. by cross-compiling), and publishes the binaries for download somewhere.   That could of course be also an opportunity for the OMP to get involved ...

**Comment by neman** [comment]
> If there is interest in the project in working on this port, it might be a good Open Mainframe Project Mentorship project...
> 
> https://www.openmainframeproject.org/projects/mentorship-program
> 
> Let me know if there is interest and I can help get this posted.

You should check this out https://youtu.be/SMF80uJuv0s

**Comment by AaronRobinsonMSFT** [comment]
@uweigand Is this issue still relevant now that s390x is supported? This issue seems misleading given the title and description.  Some comments above indicate there are still questions about CI and distribution, which I'd argue might warrant a more targeted issue. Thoughts?

**Comment by uweigand** [comment]
@AaronRobinsonMSFT the remaining open questions are around:
- Possibly providing pre-built .NET SDK & runtime binary images for s390x somewhere, similar to the pre-built tarballs provided by Microsoft for Intel and Arm.  (These would be intended for dev/test purposes only, without any official support.)
- Out-of-the box support for .NET on s390x in Linux distributons other than RHEL (and Fedora).

As to the second point, that's up to the distributions in question - if any distro out there wants to support .NET on s390x, we'd appreciate that and would be happy to help.  (For example, there's currently an issue about Alpine support open here: https://github.com/dotnet/source-build/issues/2839)

As to the first point, we're currently investigating whether we can set up an infrastructure to provide pre-built images for dev/test purposes.  But I think that would be outside the scope of this open source project, so having an issue open here isn't likely to help with that.

In summary, I agree that this issue can probably be closed now.
